File: clients/sdk-cookbook/src/recipes/cancel-mid-turn.ts
/**
 * Recipe: cancel a turn while it is running.
 *
 * Runs against `muse-conformance serve-fixture` playing the committed golden
 * transcript `schema/msp/transcripts/cancel-mid-turn` — the canned stdio host,
 * so this recipe is deterministic and needs no credentials and no model.
 *
 * What it teaches (and what the docs page walks through):
 *  - `turn/cancel` is a REQUEST, not an outcome. The ack says the host will
 *    try; the turn is over only when `turn/completed` arrives with the
 *    terminal `"cancelled"`.
 *  - a cancelled turn keeps a balanced history: the in-flight toolCall item
 *    is completed as `"cancelled"`, never dropped.
 *
 * Fixture mechanics, so the code below reads honestly: serve-fixture matches
 * our frames structurally against the transcript's client lines and then
 * replays the recorded server lines byte-exactly. Our request ids unify (the
 * SDK's default minting starts at 1, where the transcript starts), but acks
 * echo the RECORDED commandId — and `Connection.command` verifies that echo —
 * so each command below passes the transcript's commandId explicitly. Values
 * we learn from the wire (sessionId, turnId) are used as received, exactly as
 * a real client uses them.
 */

import { Host, equals, objectAt, requireHost, stringAt } from "../kit/host.js";
import { runJourney } from "../kit/segments.js";
import type { JourneyReport, Segment } from "../kit/segments.js";
import type { Recipe, RecipeHosts } from "../runner.js";

const HANDSHAKE_BUDGET_MS = 30_000;
const COMMAND_BUDGET_MS = 30_000;
const CLOSE_BUDGET_MS = 30_000;

/** The transcript's own client-side values; see the module comment. */
const WORKSPACE_ROOT = "/home/me/src/proj";
const PROMPT = "Run the flaky integration suite";
const SESSION_START_COMMAND_ID = "0198f0ab-9999-7000-8000-0000000000c1";
const TURN_START_COMMAND_ID = "018f6a1e-9b3c-7c21-a54a-2f30bd3c9f10";
const TURN_CANCEL_COMMAND_ID = "018f6a21-0f0f-7aaa-bbbb-0123456789ab";

interface Context {
  readonly conformanceBin: string;
  readonly transcript: string;
  host?: Host;
  sessionId?: string;
  turnId?: string;
}

const SEGMENTS: ReadonlyArray<Segment<Context>> = [
  {
    id: "spawn-fixture",
    title: "Spawn the canned host playing the cancel-mid-turn transcript",
    async run(context) {
      context.host = await Host.spawn(
        {
          command: context.conformanceBin,
          args: ["serve-fixture", "--transcript", context.transcript],
          env: { PATH: process.env["PATH"] ?? "/usr/bin:/bin" },
          clientInfo: { name: "conformance", version: "0.0.0" },
        },
        HANDSHAKE_BUDGET_MS,
      );
    },
  },
  {
    id: "session",
    title: "Start the session the turn will run in",
    async run(context) {
      const host = requireHost(context);
      const result = await host.msp.connection.command(
        "session/start",
        { workspaceRoot: WORKSPACE_ROOT },
        { maxAttempts: 1, commandId: SESSION_START_COMMAND_ID },
      );
      const session = objectAt(result, "session", "session/start result");
      equals(session["status"], "idle", "a fresh session's status");
      context.sessionId = stringAt(session, "sessionId", "session/start session");
    },
  },
  {
    id: "turn-start",
    title: "Start a turn and watch its work stream in",
    async run(context) {
      const host = requireHost(context);
      const sessionId = context.sessionId;
      if (sessionId === undefined) throw new Error("`session` did not finish");
      const ack = await host.msp.connection.command(
        "turn/start",
        { sessionId, input: [{ type: "text", text: PROMPT }] },
        { maxAttempts: 1, commandId: TURN_START_COMMAND_ID },
      );
      equals(ack["status"], "accepted", "turn/start ack status");
      equals(ack["disposition"], "started", "turn/start ack disposition");
      const turnId = stringAt(ack, "turnId", "turn/start ack");
      await host.waitFor(
        "the turn/started notification",
        COMMAND_BUDGET_MS,
        (notification) =>
          notification.method === "turn/started" && notification.params["turnId"] === turnId,
      );
      // The turn is genuinely mid-work when output is still streaming: the
      // fixture's toolCall item is emitting `item/delta` frames.
      await host.waitFor(
        "a streamed item/delta while the turn runs",
        COMMAND_BUDGET_MS,
        (notification) => notification.method === "item/delta",
      );
      context.turnId = turnId;
    },
  },
  {
    id: "cancel",
    title: "Ask for the cancel, then wait for the turn's own terminal",
    async run(context) {
      const host = requireHost(context);
      const sessionId = context.sessionId;
      const turnId = context.turnId;
      if (sessionId === undefined || turnId === undefined) {
        throw new Error("`turn-start` did not finish");
      }
      const ack = await host.msp.connection.command(
        "turn/cancel",
        { sessionId, turnId },
        { maxAttempts: 1, commandId: TURN_CANCEL_COMMAND_ID },
      );
      equals(ack["status"], "accepted", "turn/cancel ack status");
      // The ack is a promise to try, not the outcome. The outcome is the
      // turn's terminal.
      const completed = await host.waitFor(
        "the cancelled turn/completed notification",
        COMMAND_BUDGET_MS,
        (notification) =>
          notification.method === "turn/completed" && notification.params["turnId"] === turnId,
      );
      equals(completed.params["terminal"], "cancelled", "the turn's terminal");
      // Balanced history: the in-flight toolCall was completed as cancelled,
      // not dropped on the floor.
      const cancelledItem = host.notifications().find(
        (notification) =>
          notification.method === "item/completed" &&
          (notification.params["item"] as Record<string, unknown> | undefined)?.["status"] ===
            "cancelled",
      );
      if (cancelledItem === undefined) {
        throw new Error("the cancelled turn left no item/completed with status \"cancelled\"");
      }
    },
  },
  {
    id: "drain",
    title: "Close stdin and let the fixture host exit cleanly",
    async run(context) {
      const host = requireHost(context);
      const exit = await host.close(CLOSE_BUDGET_MS);
      equals(exit.code, 0, "the fixture host's exit code after stdin EOF");
      context.host = undefined;
    },
  },
];

export const cancelMidTurn: Recipe = {
  id: "cancel-mid-turn",
  title: "Cancel a turn while it is running",
  docsPage: "developer-docs/src/content/docs/cookbook/cancel-a-running-turn.mdx",
  needs: ["conformanceBin"],
  async run(hosts: RecipeHosts): Promise<JourneyReport> {
    const conformanceBin = hosts.conformanceBin;
    if (conformanceBin === undefined) throw new Error("conformanceBin is required");
    const context: Context = {
      conformanceBin,
      transcript: `${hosts.transcriptRoot}/cancel-mid-turn`,
    };
    return await runJourney(SEGMENTS, context, async (owned) => {
      await owned.host?.abandon(CLOSE_BUDGET_MS);
    });
  },
};

File: python/clients/sdk-cookbook-py/src/cookbook_recipes/recipes/cancel_mid_turn.py
"""Recipe twin: cancel a turn while it is running.

Runs against ``muse-conformance serve-fixture`` playing the committed golden
transcript ``schema/msp/transcripts/cancel-mid-turn``.

What it teaches: ``turn/cancel`` is a REQUEST, not an outcome — the ack says
the host will try; the turn is over only when ``turn/completed`` arrives with
the terminal ``"cancelled"``. And a cancelled turn keeps a balanced history:
the in-flight toolCall item is completed as ``"cancelled"``, never dropped.
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path

from ..kit import (
    equals,
    Host,
    JourneyReport,
    object_at,
    require_host,
    run_journey,
    Segment,
    spawn_fixture_host,
    string_at,
    within,
)
from ..runner import Recipe, RecipeHosts

HANDSHAKE_BUDGET_MS = 30_000
COMMAND_BUDGET_MS = 30_000
CLOSE_BUDGET_MS = 30_000

# The transcript's own client-side values; see the module docstring.
WORKSPACE_ROOT = "/home/me/src/proj"
PROMPT = "Run the flaky integration suite"
SESSION_START_COMMAND_ID = "0198f0ab-9999-7000-8000-0000000000c1"
TURN_START_COMMAND_ID = "018f6a1e-9b3c-7c21-a54a-2f30bd3c9f10"
TURN_CANCEL_COMMAND_ID = "018f6a21-0f0f-7aaa-bbbb-0123456789ab"


@dataclass
class Context:
    """The recipe's own state, one segment feeding the next."""

    conformance_bin: str
    transcript: str
    host: Host | None = None
    session_id: str | None = None
    turn_id: str | None = None


async def _spawn_fixture(context: Context) -> None:
    context.host = await spawn_fixture_host(
        conformance_bin=context.conformance_bin,
        transcript_dir=Path(context.transcript),
        env={"PATH": os.environ.get("PATH", "/usr/bin:/bin")},
        client_info={"name": "conformance", "version": "0.0.0"},
        budget_ms=HANDSHAKE_BUDGET_MS,
    )


async def _session(context: Context) -> None:
    host = require_host(context.host, "spawn-fixture")
    result = await within("the session/start ack", COMMAND_BUDGET_MS, host.connection.command(
        "session/start",
        {"workspaceRoot": WORKSPACE_ROOT},
        max_attempts=1,
        command_id=SESSION_START_COMMAND_ID,
    ))
    session = object_at(result, "session", "session/start result")
    equals(session.get("status"), "idle", "a fresh session's status")
    context.session_id = string_at(session, "sessionId", "session/start session")


async def _turn_start(context: Context) -> None:
    host = require_host(context.host, "spawn-fixture")
    session_id = context.session_id
    if session_id is None:
        raise RuntimeError("`session` did not finish")
    ack = await within("the turn/start ack", COMMAND_BUDGET_MS, host.connection.command(
        "turn/start",
        {"sessionId": session_id, "input": [{"type": "text", "text": PROMPT}]},
        max_attempts=1,
        command_id=TURN_START_COMMAND_ID,
    ))
    equals(ack.get("status"), "accepted", "turn/start ack status")
    equals(ack.get("disposition"), "started", "turn/start ack disposition")
    turn_id = string_at(ack, "turnId", "turn/start ack")
    await host.wait_for(
        "the turn/started notification",
        COMMAND_BUDGET_MS,
        lambda n: n.method == "turn/started" and n.params.get("turnId") == turn_id,
    )
    # The turn is genuinely mid-work when output is still streaming: the
    # fixture's toolCall item is emitting item/delta frames.
    await host.wait_for(
        "a streamed item/delta while the turn runs",
        COMMAND_BUDGET_MS,
        lambda n: n.method == "item/delta",
    )
    context.turn_id = turn_id


async def _cancel(context: Context) -> None:
    host = require_host(context.host, "spawn-fixture")
    session_id = context.session_id
    turn_id = context.turn_id
    if session_id is None or turn_id is None:
        raise RuntimeError("`turn-start` did not finish")
    ack = await within("the turn/cancel ack", COMMAND_BUDGET_MS, host.connection.command(
        "turn/cancel",
        {"sessionId": session_id, "turnId": turn_id},
        max_attempts=1,
        command_id=TURN_CANCEL_COMMAND_ID,
    ))
    equals(ack.get("status"), "accepted", "turn/cancel ack status")
    # The ack is a promise to try, not the outcome. The outcome is the turn's
    # terminal.
    completed = await host.wait_for(
        "the cancelled turn/completed notification",
        COMMAND_BUDGET_MS,
        lambda n: n.method == "turn/completed" and n.params.get("turnId") == turn_id,
    )
    equals(completed.params.get("terminal"), "cancelled", "the turn's terminal")
    # Balanced history: the in-flight toolCall was completed as cancelled, not
    # dropped on the floor.
    for notification in host.notifications():
        if notification.method != "item/completed":
            continue
        item = notification.params.get("item")
        if isinstance(item, dict) and item.get("status") == "cancelled":
            return
    raise AssertionError(
        'the cancelled turn left no item/completed with status "cancelled"'
    )


async def _drain(context: Context) -> None:
    host = require_host(context.host, "spawn-fixture")
    exit_ = await host.close(CLOSE_BUDGET_MS)
    equals(exit_.code, 0, "the fixture host's exit code after stdin EOF")
    context.host = None


SEGMENTS: tuple[Segment[Context], ...] = (
    Segment(
        "spawn-fixture", "Spawn the canned host playing the cancel-mid-turn transcript", _spawn_fixture
    ),
    Segment("session", "Start the session the turn will run in", _session),
    Segment("turn-start", "Start a turn and watch its work stream in", _turn_start),
    Segment("cancel", "Ask for the cancel, then wait for the turn's own terminal", _cancel),
    Segment("drain", "Close stdin and let the fixture host exit cleanly", _drain),
)


async def _run(hosts: RecipeHosts) -> JourneyReport:
    if hosts.conformance_bin is None:
        raise ValueError("conformance_bin is required")
    context = Context(
        conformance_bin=hosts.conformance_bin,
        transcript=f"{hosts.transcript_root}/cancel-mid-turn",
    )

    async def teardown(owned: Context) -> None:
        if owned.host is not None:
            await owned.host.abandon(CLOSE_BUDGET_MS)

    return await run_journey(SEGMENTS, context, teardown)


RECIPE = Recipe(
    id="cancel-mid-turn",
    title="Cancel a turn while it is running",
    docs_page="developer-docs/src/content/docs/cookbook/cancel-a-running-turn.mdx",
    needs=("conformance_bin",),
    run=_run,
)
